godot-core/src/obj/script.rs

```diff
@@ -417,298 +457,255 @@ impl<'a, T: ScriptInstance> DerefMut for SiMut<'a, T> {
     }
 }
 
-mod script_instance_info {
-    use std::any::type_name;
-    use std::ffi::c_void;
-    use std::mem::ManuallyDrop;
-    use std::pin::Pin;
-
-    use godot_cell::{GdCell, RefGuard};
+// Encapsulate PtrlistContainer to help ensure safety.
+mod ptrlist_container {
+    use std::collections::HashMap;
+    use std::sync::Mutex;
 
-    use crate::builtin::{GString, StringName, Variant};
-    use crate::classes::ScriptLanguage;
-    use crate::obj::Gd;
-    use crate::private::handle_panic;
-    use crate::sys;
+    use godot_ffi as sys;
 
-    use super::{ScriptInstance, ScriptInstanceData, SiMut};
-
-    fn borrow_panic<T: ScriptInstance, R>(err: Box<dyn std::error::Error>) -> R {
-        panic!(
-            "\
-                ScriptInstance borrow failed, already bound; T = {}.\n  \
-                Make sure to use `SiMut::base_mut()` when possible.\n  \
-                Details: {err}.\
-            ",
-            type_name::<T>(),
-        );
-    }
-
-    fn borrow_instance<T: ScriptInstance>(instance: &ScriptInstanceData<T>) -> RefGuard<'_, T> {
-        instance
-            .inner
-            .as_ref()
-            .borrow()
-            .unwrap_or_else(borrow_panic::<T, _>)
-    }
-
-    fn borrow_instance_cell<T: ScriptInstance>(
-        instance: &ScriptInstanceData<T>,
-    ) -> Pin<&GdCell<T>> {
-        instance.inner.as_ref()
-    }
-
-    /// # Safety
-    ///
-    /// `p_instance` must be a valid pointer to a `ScriptInstanceData<T>` for the duration of `'a`.
-    /// This pointer must have been created by [super::create_script_instance] and transfered to Godot.
-    unsafe fn instance_data_as_script_instance<'a, T: ScriptInstance>(
-        p_instance: sys::GDExtensionScriptInstanceDataPtr,
-    ) -> &'a ScriptInstanceData<T> {
-        &*(p_instance as *mut ScriptInstanceData<T>)
-    }
-
-    fn transfer_bool_to_godot(value: bool) -> sys::GDExtensionBool {
-        value as sys::GDExtensionBool
-    }
-
-    /// # Safety
-    ///
-    /// - We expect the engine to provide a valid variant pointer the return value can be moved into.
-    unsafe fn transfer_variant_to_godot(variant: Variant, return_ptr: sys::GDExtensionVariantPtr) {
-        variant.move_into_var_ptr(return_ptr)
+    /// Helper struct to store the lengths of lists so they can be properly freed.
+    pub struct PtrlistContainer<T> {
+        list_lengths: Mutex<HashMap<*mut T, u32>>,
```

Comment:
Nice abstraction!

Re name, "list" and "container" are somewhat redundant, or it's not really clear how they relate. Maybe we should somehow express that this stores the lengths? `BoundedPtrList` or something like that?

You could also mention that you use "list" and not "array" because these refer to "property/method lists" (in Godot terms).